### DISCUSSION TITLE
nonempty is deprecated

### DISCUSSION THREAD

**Discussion by FabioZan86** [colinhacks/zod#2847] [Q&A] [answered]
Hello, what reason to make it obsolete? How am I going to display this message now that the field is mandatory?
```
password: z.string()
    .nonempty('The Password field is required')
    .min(6, 'Password must be longer than 6 characters')
    .max(8, 'The password must not be a maximum of 8 characters'),
```

**Comment by renovatt** [comment]
I have the same question, what I have to use now?

**Reply by Pizle0210** [reply]
Just make sure it has a .min() which covers for .nonempty().

**Comment by samchungy** [comment, marked as answer]
I'm confused by your example, doesn't `.min(6)` cover your `.nonempty` case? I would assume Password must be longer than 6 characters covers that it's required. `.min(1)` would cover `.nonempty()` in other cases

**Reply by iway1** [reply]
No it doesn't because the error message is different, `.min(1)` would work though. It's a bit subjective but some would argue we should be showing a separate error message when empty versus not enough characters, pretty common I would say

**Comment by FabioZan86** [comment]
@samchungy I didn't understand what you meant. I just want to inform you that the field is mandatory. I don't want to inform you that the minimum is 6 to be mandatory. Every time the lib is being updated it becomes difficult to stay secure. I don't know why 'not empty' became obsolete!

**Reply by lywoo-sroeng** [reply]
@FabioZan86 Based on [the official doc](https://zod.dev/?id=strings), you achieve that by implementing the following code.
```
password: z
          .string({
            required_error: 'The Password field is required',
          })
          .min(6, 'Password must be longer than 6 characters')
          .max(8, 'The password must not be a maximum of 8 characters'),
```

**Reply by FabioZan86** [reply]
@lywoo-sroeng I want to thank you for your help, but it doesn't work! The error that warns when submitting the form is: 'The password must be more than 6 characters' and not 'The Password field is required'. I don't know why something so simple is complicated. Taking advantage, I also have a problem here: https://codesandbox.io/s/hook-form-zod-mui-forked-4wjqv6 it says: message: "Expected boolean, string received"
type: "invalid_type"'.

**Reply by samchungy** [reply]
You should be able to also do: `.min(1, 'The Password field is required).min(6, 'Password must be longer than 6 characters')`. Though this seems fairly redundant imo.

**Reply by FabioZan86** [reply]
@samchungy Thank you very much! It worked perfectly for my case. I hadn't thought about that perspective, it may seem redundant, but it worked! The other 'codesandbox' example was missing z.coerce. Already ok.

**Reply by idlethumbs** [reply]
I'm using i18n translations for my errors, which means i have error mapping which previously had `invalid_type_received_undefined` mapped to 'field is required', this worked with non-empty, 
using @samchungy's solution would work in situ but i would have to manually apply translations per zod schema in order to ensure that 'field is required' would be translated, whereas before it would apply globally,
not sure what the best solution is as 'field should be at least 1 character(s)' while accurate is a bit strange (and long!) from a UX point of view for a required field error.

**Reply by idlethumbs** [reply]
to answer my own comment - it seems that  omitting min() and nonEmpty() - e.g. `firstName: z.string()` still results in a required field, and as long as i don't add .optional() to the field i can get the original required error to trigger.

**Reply by almeidapaulooliveira** [reply]
Use Yup !

**Reply by iway1** [reply]
@idlethumbs z.string() doesn't result in a required field because it will accept the empty string

**Comment by dacmarcell** [comment]
What about strings? Which method would assume `nonempty`?

**Reply by netojose** [reply]
> What about strings? Which method would assume `nonempty`?

```ts
min(1)
```

**Reply by dacmarcell** [reply]
Tks!

**Reply by netojose** [reply]
You're welcome, @marcelldac

**Reply by oscarvedo** [reply]
> > What about strings? Which method would assume `nonempty`?
> 
> ```ts
> min(1)
> ```

Strings filled with spaces will pass the checks.

**Reply by jeremy-habit** [reply]
.trim().min(1)

**Comment by SorooshGb** [comment]
Why aren't the people feeling smart by answering "just do .min(1)" see how stupid this is?! 
There should be a dedicated method for checking empty values and a separate one for minimum length, end of story. Doing .min(1, "Field is required") is not only unintuitive but also stupid. No validation or form library operates like this; empty-value validation and minimum-length validation are always separate concerns.

**Reply by samchungy** [reply]
>Why aren't the people feeling smart by answering "just do .min(1)" see how stupid this is?!

My advice and opinions are free and apparently live rent free in some minds 😆.Take it or leave it 🤷

ps. JSON Schema does this 🤷

**Comment by kylecarhart** [comment]
For those coming to this discussion from google, `nonempty` for string has been undeprecated as of 0c6cbbdd1315683dd3d589fbdc5765c26431dcc9.

**Comment by iway1** [comment]
the deprecation here seems kind of random and counterproductive. 

It's really subjective whether or not `nonempty()` is better than `.min(1)`, both work fine and have identical behavior. But since it is subjective and already implemented, why not just leave it in and make upgrading versions easier for people who already use `.nonempty()`?

Edit: Oh lol just saw it's undeprecated now, good call

**Comment by Anoesj** [comment]
Why was `z.string().nonempty()` not updated to be the equivalent of `z.string().trim().min(1)`?

**Reply by NuclEnergy** [reply]
What if spaces could also be considered valid input?

**Comment by HideoKun** [comment]
@FabioZan86 - use this as a starter

```
const zString = (min: number = 2, max: number = 100) =>
  z.string()
    .trim()
    .min(min)
    .max(max);
```
